File: reverse_analysis/ghidra_extract.java
// Ghidra script: extract cryptngc.dll function list and BCrypt/NCrypt references
//@category Analysis

import ghidra.app.script.GhidraScript;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.*;
import ghidra.program.model.address.*;
import ghidra.program.model.lang.*;
import ghidra.program.model.pcode.*;
import ghidra.app.decompiler.*;
import ghidra.util.task.TaskMonitor;

public class ghidra_extract extends GhidraScript {

    @Override
    public void run() throws Exception {
        Listing listing = currentProgram.getListing();
        FunctionManager fm = currentProgram.getFunctionManager();
        ReferenceManager refMgr = currentProgram.getReferenceManager();
        ExternalManager extMgr = currentProgram.getExternalManager();

        // ─── 1. List all functions ───────────────────────────
        println("=" + repeat("=", 78) + "=");
        println("ALL FUNCTIONS IN cryptngc.dll");
        println("=" + repeat("=", 78) + "=");

        FunctionIterator funcs = fm.getFunctions(true);
        int count = 0;
        while (funcs.hasNext()) {
            Function f = funcs.next();
            if (!f.isThunk()) {
                println(f.getEntryPoint().toString(false) + "  " + f.getName());
                count++;
            }
        }
        println("\nTotal non-thunk functions: " + count);

        // ─── 2. Find BCrypt* / NCrypt* external functions and their callers ──
        println("\n" + "=" + repeat("=", 78) + "=");
        println("BCRYPT/NCRYPT IMPORT CALL SITES");
        println("=" + repeat("=", 78) + "=");

        for (ExternalLocation extLoc : extMgr.getExternalLocations()) {
            String libName = extLoc.getLibraryName();
            String funcName = extLoc.getFunctionName();
            if (funcName == null) continue;
            if (!libName.toLowerCase().contains("bcrypt") &&
                !libName.toLowerCase().contains("ncrypt")) continue;

            Address extAddr = extLoc.getExternalSpaceAddress();
            if (extAddr == null) continue;

            ReferenceIterator refs = refMgr.getReferencesTo(extAddr);
            boolean hasRefs = false;
            while (refs.hasNext()) {
                Reference ref = refs.next();
                if (ref.getReferenceType().isCall()) {
                    if (!hasRefs) {
                        println("\n" + funcName + " [" + libName + "] @ " + extAddr);
                        hasRefs = true;
                    }
                    Address fromAddr = ref.getFromAddress();
                    Function caller = fm.getFunctionContaining(fromAddr);
                    if (caller != null) {
                        println("    CALL <- " + caller.getName() + " @ " + caller.getEntryPoint());
                    } else {
                        println("    CALL <- " + fromAddr + " (no function)");
                    }
                }
            }
        }

        // ─── 3. Decompile key functions ──────────────────────
        println("\n" + "=" + repeat("=", 78) + "=");
        println("DECOMPILED BCrypt/NCrypt CALLS IN KEY FUNCTIONS");
        println("=" + repeat("=", 78) + "=");

        DecompInterface decomp = new DecompInterface();
        decomp.setOptions(new DecompileOptions());
        decomp.openProgram(currentProgram);

        String[] keyFuncs = {
            "NgcDecryptWithUserIdKey", "NgcDecryptWithUserIdKeySilent",
            "NgcCreateUserIdKey", "NgcCreateUserIdKeyEx", "NgcCreateUserIdKeyHandle",
            "NgcSignWithUserIdKey", "NgcSignWithUserIdKeySilent",
            "NgcChangePin", "NgcChangePinSilent",
            "NgcCreateContainer", "NgcCreateContainerSilent",
            "NgcOpenUserIdKey", "NgcGetUserIdKeyName",
            "NgcPackAuthBuffer", "NgcUnpackAuthBuffer",
            "NgcEncryptWithAsymmetricKey", "FidoCreateCredential"
        };

        funcs = fm.getFunctions(true);
        while (funcs.hasNext()) {
            Function func = funcs.next();
            String fname = func.getName();

            // Check if this is a key function
            boolean isKey = false;
            for (String kf : keyFuncs) {
                if (fname.contains(kf) || kf.contains(fname)) {
                    isKey = true;
                    break;
                }
            }
            if (!isKey) continue;

            println("\n--- " + fname + " @ " + func.getEntryPoint() + " ---");

            DecompileResults res = decomp.decompileFunction(func, 30, monitor);
            if (res == null || !res.decompileCompleted()) {
                println("  Decompile failed: " + res.getErrorMessage());
                continue;
            }

            HighFunction hf = res.getHighFunction();
            if (hf == null) {
                println("  No high function");
                continue;
            }

            // Walk pcode to find CALL operations
            Iterator<PcodeOpAST> pcodeIter = hf.getPcodeOps();
            while (pcodeIter.hasNext()) {
                PcodeOpAST op = pcodeIter.next();
                int opcode = op.getOpcode();

                if (opcode == PcodeOp.CALL || opcode == PcodeOp.CALLIND) {
                    Varnode[] inputs = op.getInputs();
                    if (inputs != null && inputs.length > 0) {
                        Varnode target = inputs[0];
                        Address targetAddr = target.getAddress();
                        Symbol sym = getSymbolAt(targetAddr);
                        if (sym != null) {
                            String symName = sym.getName();
                            if (symName.contains("BCrypt") || symName.contains("NCrypt") ||
                                symName.contains("CryptProtect") || symName.contains("CryptUnprotect") ||
                                symName.contains("RpcBinding") || symName.contains("NdrClientCall")) {
                                println("    " + op.toString());
                            }
                        }
                    }
                }
            }
        }

        println("\n" + "ANALYSIS COMPLETE");
    }

    private String repeat(String s, int n) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < n; i++) sb.append(s);
        return sb.toString();
    }
}

File: reverse_analysis/ghidra_extract.py
# Ghidra script: extract cryptngc.dll function list and BCrypt/NCrypt call sites
# Run as: analyzeHeadless <project> <project_name> -process cryptngc.dll -postScript ghidra_extract.py

from ghidra.program.model.listing import Function
from ghidra.program.model.symbol import SourceType
from ghidra.program.model.lang import OperandType
from ghidra.program.model.address import Address
from ghidra.program.model.symbol import RefType
from ghidra.program.model.block import SimpleBlockModel, CodeBlock
from ghidra.util.task import ConsoleTaskMonitor
from ghidra.app.decompiler import DecompInterface, DecompileOptions
from ghidra.program.model.pcode import PcodeOp
import jarray

monitor = ConsoleTaskMonitor()

# ────────────────────────────────────────────────────────────
# 1. List all functions
# ────────────────────────────────────────────────────────────
print("=" * 80)
print("ALL FUNCTIONS IN cryptngc.dll")
print("=" * 80)

fm = currentProgram.getFunctionManager()
functions = fm.getFunctions(True)

for func in functions:
    body = func.getBody()
    name = func.getName()
    addr = func.getEntryPoint()
    size = body.getNumAddresses()
    # count instructions
    code_units = []
    code_manager = currentProgram.getCodeManager()
    cu = code_manager.getCodeUnitAt(addr)
    if cu:
        print(f"0x{addr.toString(False):>16}  {name}")

# ────────────────────────────────────────────────────────────
# 2. Find functions that reference BCrypt*/NCrypt* imports
# ────────────────────────────────────────────────────────────
print()
print("=" * 80)
print("FUNCTIONS CALLING BCrypt or NCrypt APIs")
print("=" * 80)

# Get the external manager
ext_mgr = currentProgram.getExternalManager()
ext_list = ext_mgr.getExternalLibraries()
crypto_libs = []

for lib in ext_list:
    lib_name = lib.getName().lower()
    if 'bcrypt' in lib_name or 'ncrypt' in lib_name:
        crypto_libs.append(lib)
        print(f"\nLibrary: {lib.getName()}")

# For each external function in bcrypt/ncrypt
for lib in crypto_libs:
    for ext_func in lib.getExternalFunctions():
        func_name = ext_func.getName()
        ext_addr = ext_func.getExternalLocationAddress()

        # Find all references to this external function
        ref_mgr = currentProgram.getReferenceManager()
        refs = ref_mgr.getReferencesTo(ext_addr)

        callers = set()
        for ref in refs:
            if ref.getReferenceType().isCall():
                caller_addr = ref.getFromAddress()
                caller_func = fm.getFunctionContaining(caller_addr)
                if caller_func:
                    callers.add(caller_func.getName())

        if callers:
            print(f"\n  {func_name} @ {ext_addr.toString(False)}")
            for caller in sorted(callers):
                print(f"    <- {caller}")

# ────────────────────────────────────────────────────────────
# 3. Decompile key functions and search for crypto calls
# ────────────────────────────────────────────────────────────
print()
print("=" * 80)
print("DECOMPILING KEY FUNCTIONS")
print("=" * 80)

# Initialize decompiler
iface = DecompInterface()
iface.setOptions(DecompileOptions())
iface.openProgram(currentProgram)

key_func_names = [
    "NgcDecryptWithUserIdKey",
    "NgcDecryptWithUserIdKeySilent",
    "NgcCreateUserIdKey",
    "NgcCreateUserIdKeyEx",
    "NgcCreateUserIdKeyHandle",
    "NgcSignWithUserIdKey",
    "NgcSignWithUserIdKeySilent",
    "NgcChangePin",
    "NgcChangePinSilent",
    "NgcCreateContainer",
    "NgcOpenUserIdKey",
    "FUN_1800197F0",  # NgcCreateUserIdKeyHandle alternate
]

for func in functions:
    name = func.getName()
    for kn in key_func_names:
        if kn in name or name in kn:
            print(f"\n--- {name} @ {func.getEntryPoint()} ---")
            res = iface.decompileFunction(func, 30, monitor)
            if res and res.getHighFunction():
                hf = res.getHighFunction()
                pcode = hf.getPcodeOps()
                while pcode.hasNext():
                    op = pcode.next()
                    mnem = op.getOpcode()
                    # Check for CALL operations
                    if mnem == PcodeOp.CALL:
                        inputs = op.getInputs()
                        if inputs and len(inputs) > 0:
                            target = inputs[0]
                            print(f"  CALL: {op}")
                    elif mnem == PcodeOp.CALLIND:
                        print(f"  CALLIND: {op}")
            break

print()
print("ANALYSIS COMPLETE")
